bigquery/google/cloud/bigquery/_helpers.py

```diff
@@ -28,6 +28,11 @@
 
 _RFC3339_MICROS_NO_ZULU = '%Y-%m-%dT%H:%M:%S.%f'
 
+# Canonical formats for timestamps in BigQuery
+# g.co/cloud/bigquery/docs/reference/standard-sql/data-types#timestamp-type
+_BIGQUERY_TIMESTAMP = '%Y-%m-%d %H:%M:%S%z'
+_BIGQUERY_TIMESTAMP_MICROS = '%Y-%m-%d %H:%M:%S.%f%z'
```

Comment:
`%z` is one of those supported-by-`strftime`-but-not-`strptime` directives in Python2. :(

If new the zone would always be UTC, we could just strip off the `+00:00` and add the tzinfo ourselves.  As it is, the spec makes it super-complicated:  not only to we have to parse the offset, but we also have to be prepared to deal with zone names.

Reply:
@tswast unless, of course, you can assure us that the back-end will normalize everything to UTC and always send it back as `+00:00`. :)

Reply:
True, I'll check with the backend folks to see if this will always be UTC when coming from the API.

Reply:
This passes on 2.7 and 3.x, but we have a coverage gap (any non-null value with non `+00:00` zone):

```python
--- a/bigquery/google/cloud/bigquery/_helpers.py
+++ b/bigquery/google/cloud/bigquery/_helpers.py
@@ -30,8 +30,11 @@ _RFC3339_MICROS_NO_ZULU = '%Y-%m-%dT%H:%M:%S.%f'
 
 # Canonical formats for timestamps in BigQuery
 # g.co/cloud/bigquery/docs/reference/standard-sql/data-types#timestamp-type
+_BIGQUERY_TIMESTAMP_NO_OFFSET = '%Y-%m-%d %H:%M:%S'
 _BIGQUERY_TIMESTAMP = '%Y-%m-%d %H:%M:%S%z'
+_BIGQUERY_TIMESTAMP_MICROS_NO_OFFSET = '%Y-%m-%d %H:%M:%S.%f'
 _BIGQUERY_TIMESTAMP_MICROS = '%Y-%m-%d %H:%M:%S.%f%z'
+_ZULU_NO_OFFSET = '+00:00'
 
 
 def _not_null(value, field):
@@ -88,16 +91,17 @@ def _timestamp_query_param_from_json(value, field):
         :data:`None`).
     """
     if _not_null(value, field):
-        # Remove final : from timezone because python represents UTC offset
-        # as +HHMM or -HHMM, whereas BigQuery represents it as +HH:MM or -HH:MM
-        value = ''.join(value.rsplit(':', 1))
-        if '.' in value:
-            # YYYY-MM-DDTHH:MM:SS.ffffff+00:00
-            return datetime.datetime.strptime(
-                value, _BIGQUERY_TIMESTAMP_MICROS)
-        else:
-            # YYYY-MM-DDTHH:MM:SS+00:00
-            return datetime.datetime.strptime(value, _BIGQUERY_TIMESTAMP)
+
+        if value.endswith(_ZULU_NO_OFFSET):
+            value = value[:-len(_ZULU_NO_OFFSET)]
+            if '.' in value:
+                return datetime.datetime.strptime(
+                    value, _BIGQUERY_TIMESTAMP_MICROS_NO_OFFSET).replace(
+                        tzinfo=UTC)
+            else:
+                return datetime.datetime.strptime(
+                    value, _BIGQUERY_TIMESTAMP_NO_OFFSET).replace(
+                        tzinfo=UTC)
     else:
         return None
 
```

Reply:
I got confirmation that the timestamp will always be encoded as UTC.

The internal protos say it'll have trailing Z instead of +00:00, though, so I'll do a little more digging.

Reply:
I've pushed a new commit. There are 8 possible formats that the BigQuery API might give us that we'll have to handle.

- 'T' vs ' ' between time and date.
- 'Z' vs '+00:00' for representing UTC time.
- microseconds or no microseconds.

My solution is to just try them all, but maybe we should be constructing the time format string on the fly instead?